### ISSUE TITLE
question about evaluation metric code

### ISSUE THREAD

**Issue by Butterfies** [cps11/catvrnn#1]
Hi @cps11，Thanks for your work. While reading your paper, I noticed that you mentioned some metrics, including Acc, BLEU, and PPL, but didn't see the relevant code in the repository, and I wanted to take a deeper look at how these metrics are calculated and the use of tools/CNNClassifier .py. I would appreciate it if you would like to share relevant code or other resources. 
Besides，I have a question that only the training set in the dataset without the test set, the reference object of the bleu metric is the training set, whether it is reasonable.

**Comment by cps11** [comment]
The calculation function for "ACC" is as follows.

```
def get_accuracy(model, data: torch.Tensor, device: torch.Tensor) -> float:
    total_acc = 0
    labels = torch.tensor([0 for _ in range(len(data))], device=device)

    batch_size = 128
    data_size = len(data)
    max_iters = math.ceil(data_size / batch_size)

    for i in range(max_iters):
        offset = i * batch_size 

        pred = model.fit(data[offset:offset+batch_size])
        total_acc += torch.sum((pred == labels[offset:offset+batch_size])).item()
    
    return total_acc/data_size
```
    
The calculation function for "PPL" is as follows.  "probabilities" represents the sampling probability for each word during the generation process.

```
def calculate_perplexity(probabilities):
    N = len(probabilities)  
    log_sum = sum(math.log2(p) for p in probabilities)  
    perplexity = 2 ** (-log_sum / N)  
    return perplexity
```

We have utilized an open-source Python library for evaluating the quality of generated text, and the code can be found at this address: https://github.com/Danial-Alh/fast-bleu. The calculation for forward BLEU and backward BLEU involves only swapping the positions of the reference text and generated text parameters. We believe you can easily implement this functionality based on the research paper.

It is reasonable to use the training set as the reference object for the BLEU metric in text generation tasks. This is because text generation tasks involve learning the probability distribution of the training set texts. By using the training set as the reference, we can more effectively evaluate how well the generated text matches the patterns and language usage present in the training data. Using a separate test set as the reference may introduce biases or discrepancies that do not accurately reflect the model's ability to generalize from the training data. Therefore, using the training set as a reference provides a more reliable measure of the model's performance in generating texts.

**Comment by Butterfies** [comment]
> The calculation function for "ACC" is as follows.
> 
> ```
> def get_accuracy(model, data: torch.Tensor, device: torch.Tensor) -> float:
>     total_acc = 0
>     labels = torch.tensor([0 for _ in range(len(data))], device=device)
> 
>     batch_size = 128
>     data_size = len(data)
>     max_iters = math.ceil(data_size / batch_size)
> 
>     for i in range(max_iters):
>         offset = i * batch_size 
> 
>         pred = model.fit(data[offset:offset+batch_size])
>         total_acc += torch.sum((pred == labels[offset:offset+batch_size])).item()
>     
>     return total_acc/data_size
> ```
> 
> The calculation function for "PPL" is as follows. "probabilities" represents the sampling probability for each word during the generation process.
> 
> ```
> def calculate_perplexity(probabilities):
>     N = len(probabilities)  
>     log_sum = sum(math.log2(p) for p in probabilities)  
>     perplexity = 2 ** (-log_sum / N)  
>     return perplexity
> ```
> 
> We have utilized an open-source Python library for evaluating the quality of generated text, and the code can be found at this address: https://github.com/Danial-Alh/fast-bleu. The calculation for forward BLEU and backward BLEU involves only swapping the positions of the reference text and generated text parameters. We believe you can easily implement this functionality based on the research paper.
> 
> It is reasonable to use the training set as the reference object for the BLEU metric in text generation tasks. This is because text generation tasks involve learning the probability distribution of the training set texts. By using the training set as the reference, we can more effectively evaluate how well the generated text matches the patterns and language usage present in the training data. Using a separate test set as the reference may introduce biases or discrepancies that do not accurately reflect the model's ability to generalize from the training data. Therefore, using the training set as a reference provides a more reliable measure of the model's performance in generating texts.

Thank you for your reply!
